File: algorithms/python/ring_buffer.py
# This file implements ring buffer which is usually used in embedded systems for handling UART data etc.
# See wiki to learn more about the concept of ring buffer
import unittest

class CircularBuffer:
  
  def __init__(self, size):
    self._buffer = [None]
    self._size = size
    
    if isinstance(size, int) and size > 0:
      self._buffer *= self._size
    else:
      self._buffer *= 10
      print("Please ensure value is non-negative and unsigned integer. Defaulting to buffer size of 10!")

    self._head_ptr = 0
    self._tail_ptr = 0
    self._buff_curr_size = 0
    self._is_full = False

  def movePointerForward(self, ptr_type):
    if ptr_type == "h":
      self._head_ptr = (self._head_ptr + 1) % self._size
    elif ptr_type == "t":
      self._tail_ptr = (self._tail_ptr + 1) % self._size
    else:
      raise SyntaxError("Invalid Pointer Type. Only head[h] or tail[t] type valid")
    pass
  
  def isBufferFull(self):
    # Buffer is full when the head's position equals the tail's position
    return self._is_full

  def isBufferEmpty(self):
    if self._head_ptr == self._tail_ptr and not self._is_full:
      return True
    else:
      return False

  def writeData(self, new_data):
    # Update the new data at the head's position if not full
    # Otherwise, if full, then the tail will move forward and the data will be overwritten
    if self.isBufferFull():
      self.movePointerForward("t")

    self._buffer[self._head_ptr] = new_data
    self.movePointerForward("h")

    if self._head_ptr == self._tail_ptr:
      self._is_full = True

  def readData(self):

    if self.isBufferEmpty():
      raise Exception("Buffer is currently empty!")

    # Read the data from the current tail position and then move a step forward
    data = self._buffer[self._tail_ptr]
    self.movePointerForward("t")
    
    # Since we have read the data, the buffer cannot be full
    self._is_full = False

    return data

  def displayBufferData(self):
    msg = ""

    for i in range(self._size):
      msg += str(self._buffer[i]) + " "
      if self._head_ptr == i:
        msg += " [h] "
      if self._tail_ptr == i:
        msg += " [t] "

      msg += " | "
    
    print(msg)

class TestCircularBuffer(unittest.TestCase):

  def create_buffer(self, size):
    return CircularBuffer(size)

  def test_small_buffer_write_read(self):
    n = 5
    test_buffer = self.create_buffer(n)
    
    # Buffer should not be full or empty during initialization
    self.assertEqual(test_buffer.isBufferEmpty(), True)
    test_buffer.isBufferFull()
    self.assertEqual(test_buffer._is_full, False)

    # Populate some data into the buffer to make it full
    for i in range(n):
      test_buffer.writeData(10 + i)
    
    test_buffer.displayBufferData()
    
    test_buffer.isBufferFull()
    self.assertEqual(test_buffer._is_full, True)

    test_buffer.readData()
    self.assertEqual(test_buffer._tail_ptr, 1)
    self.assertEqual(test_buffer._head_ptr, 0)
    self.assertEqual(test_buffer._is_full, False)
    test_buffer.displayBufferData()

    test_buffer.writeData(20)
    self.assertEqual(test_buffer._head_ptr, 1)
    self.assertEqual(test_buffer._tail_ptr, 1)
    self.assertEqual(test_buffer._is_full, True)
    test_buffer.displayBufferData()

if __name__ == '__main__':
  unittest.main()

File: algorithms/cpp/src/ring_buffer.cpp

#include "ring_buffer.hpp"  // Include the CircularBuffer class

template <class T, std::size_t Size>
algorithms::CircularBuffer<T, Size>::CircularBuffer() :
  _head_idx(0), _tail_idx(0), _is_full(false), _buffer {} {
    std::cout << "Circular Buffer initialized!" << std::endl;
}

template <class T, std::size_t Size>
bool algorithms::CircularBuffer<T, Size>::isBufferEmpty() {
  bool is_empty;
  if (this->_head_idx == this->_tail_idx && not this->_is_full) {
    is_empty = true;
  } else {
    is_empty = false;
  }
  return is_empty;
}

template <class T, std::size_t Size>
bool algorithms::CircularBuffer<T, Size>::isBufferFull() {
  return _is_full;
}

template <class T, std::size_t Size>
void algorithms::CircularBuffer<T, Size>::writeData(const T &new_data) {
  // Need to check if the buffer is full first

  if (this->isBufferFull()) {
    // If buffer is full, then the tail index will increment one step forward, and 
    // data will then be overwritten with new incoming data
    this->_tail_idx += (this->_tail_idx + 1) % Size;
  }

  // Now, we can use the head index to overwrite the data if the buffer was full
  // When writing new data, the data will be placed at wherever the head index is
  // and then it will increment by one step
  this->_buffer[_head_idx] = new_data;
  this->_head_idx += (this->_head_idx + 1) % Size;

  if (this->_head_idx == this->_tail_idx) {
    this->_is_full = true;
  }
}

template <class T, std::size_t Size>
T algorithms::CircularBuffer<T, Size>::readData() {
  // Get the data from where the tail index is currently pointing at
  T data = this->_buffer[_tail_idx];
  this->_tail_idx += (this->_tail_idx + 1) % Size;
  return data;
}

template <class T, std::size_t Size>
std::size_t algorithms::CircularBuffer<T, Size>::getHeadIdx() {
  return _head_idx;
}

int main() {
  algorithms::CircularBuffer<int, 5> circular_buffer_1;
  std::cout << circular_buffer_1.isBufferEmpty() << std::endl;
  int test_data = 3;
  circular_buffer_1.writeData(test_data);
  std::cout << circular_buffer_1.isBufferEmpty() << std::endl;
  return 0;
}